In [1]:
import pandas as pd
import numpy as np

In [2]:
source_dir = '../data/undesired_outcomes_pred/'
dest_dir = source_dir
extension = '.csv'

In [3]:
elog = pd.read_csv(source_dir.__add__('elog.csv'))

In [4]:
pd.options.display.max_columns = 100

In [5]:
elog.sample(7)

Case ID                                        Activity  \
232102   c7900278faf29be2          Payment application-Application-decide   
2112204  2107e3df1a62d95e     Geo parcel document-Reported-finish editing   
2381742  eba60dd715d14930                    Inspection-Remote-initialize   
200069   19c8ba801a27f64d            Parcel document-Main-insert document   
532477   541dce78d4d3caa1              Parcel document-Main-begin editing   
1088917  0094067780007c66  Payment application-Application-finish editing   
1351307  d2fb9e04f0b60c31          Geo parcel document-Declared-calculate   

                              Resource       Complete Timestamp  \
232102                          8beb64  2015-12-10 16:58:31.518   
2112204                         727350  2017-12-03 15:14:48.789   
2381742           Inspection automaton  2017-06-04 12:35:26.571   
200069                          8a1fba  2015-09-11 12:32:27.000   
532477   Document processing automaton  2015-07-07 08:03:06.000   
1088917                         727350  2016-12-02 08:44:07.740   
1351307                         019209  2016-11-14 12:05:34.461   

         (case) amount_applied0  (case) amount_applied1  \
232102                 10602.96                     NaN   
2112204                 4688.47                     NaN   
2381742                31092.25                     NaN   
200069                  9868.40                     NaN   
532477                 43403.12                     NaN   
1088917                 3926.42                     NaN   
1351307                30899.94                30899.94   

         (case) amount_applied2  (case) amount_applied3  (case) applicant  \
232102                      NaN                     NaN  01fb765451bb8046   
2112204                     NaN                     NaN  ef396b14e592629d   
2381742                     NaN                     NaN  f32bc589cf06c637   
200069                      NaN                     NaN  c0626a1d67f2f7dd   
532477                      NaN                     NaN  3e0bd8469dda1e2d   
1088917                     NaN                     NaN  fb5e63ff5128f24f   
1351307                30868.18                     NaN  8400be89615659fe   

         (case) area  (case) basic payment  (case) cross_compliance  \
232102       31.0560                  True                      0.0   
2112204      13.9695                  True                      0.0   
2381742     105.9568                  True                      0.0   
200069       28.7111                  True                      0.0   
532477      144.1383                  True                      0.0   
1088917      11.0825                  True                      0.0   
1351307     103.6659                  True                      0.0   

        (case) department  (case) greening  (case) number_parcels  \
232102                 4e             True                     12   
2112204                4e             True                      8   
2381742                6b             True                     20   
200069                 e7             True                      5   
532477                 4e             True                     44   
1088917                d4             True                      3   
1351307                4e             True                     59   

         (case) payment_actual0  (case) payment_actual1  \
232102                 10572.41                     NaN   
2112204                 4687.64                     NaN   
2381742                31082.71                     NaN   
200069                  9843.65                     NaN   
532477                 43388.56                     NaN   
1088917                 3927.17                     NaN   
1351307                30808.54                     0.0   

         (case) payment_actual2  (case) payment_actual3  (case) penalty_ABP  \
232102                      NaN                     NaN               False   
2112204              

In [6]:
elog.loc[~pd.isnull(elog['(case) penalty_amount2'])]

Case ID                                           Activity  \
8383     b0d199e3fd21ffb4        Payment application-Application-mail income   
8384     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8385     b0d199e3fd21ffb4         Payment application-Application-mail valid   
8386     b0d199e3fd21ffb4            Entitlement application-Main-mail valid   
8387     b0d199e3fd21ffb4                    Control summary-Main-initialize   
8388     b0d199e3fd21ffb4                 Control summary-Main-begin editing   
8389     b0d199e3fd21ffb4                Control summary-Main-finish editing   
8390     b0d199e3fd21ffb4                    Parcel document-Main-initialize   
8391     b0d199e3fd21ffb4                 Parcel document-Main-begin editing   
8392     b0d199e3fd21ffb4               Parcel document-Main-insert document   
8393     b0d199e3fd21ffb4                          Parcel document-Main-save   
8394     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8395     b0d199e3fd21ffb4                Reference alignment-Main-initialize   
8396     b0d199e3fd21ffb4                 Reference alignment-Main-performed   
8397     b0d199e3fd21ffb4          Department control parcels-Main-performed   
8398     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8399     b0d199e3fd21ffb4            Entitlement application-Main-initialize   
8400     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8401     b0d199e3fd21ffb4                          Parcel document-Main-save   
8402     b0d199e3fd21ffb4                Parcel document-Main-finish editing   
8403     b0d199e3fd21ffb4         Entitlement application-Main-begin editing   
8404     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8405     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8406     b0d199e3fd21ffb4                      Reference alignment-Main-save   
8407     b0d199e3fd21ffb4      Department control parcels-Main-begin editing   
8408     b0d199e3fd21ffb4         Payment application-Application-initialize   
8409     b0d199e3fd21ffb4      Payment application-Application-begin editing   
8410     b0d199e3fd21ffb4          Payment application-Application-calculate   
8411     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
8412     b0d199e3fd21ffb4             Entitlement application-Main-calculate   
...                   ...                                                ...   
1641626  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641627  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641628  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641629  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641630  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641631  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641632  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641633  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641634  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641635  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641636  f24f9e1ac0d7701f      Payment application-Objection-revoke approval   
1641637  f24f9e1ac0d7701f  Payment application-Objection-begin admissibil...   
1641638  f24f9e1ac0d7701f  Payment application-Objection-check admissibility   
1641639  f24f9e1ac0d7701f            Payment application-Objection-calculate   
1641640  f24f9e1ac0d7701f               Payment application-Objection-refuse   
1641641  f24f9e1ac0d7701f  Payment application-Objection-begin editing fr...   
1641642  f24f9e1ac0d7701f      Payment application-Objection-insert document   
1641643  f24f9e1ac0d7701f                      Payment application-Main-save   
1641644  f24f9e1ac

In [7]:
elog.loc[~pd.isnull(elog['(case) payment_actual3'])].sample(3)

Case ID                             Activity  \
143184   4a9acc5059b9f46a            Parcel document-Main-save   
1494673  6dcac1562d4ca766  Geo parcel document-Main-initialize   
1495362  6dcac1562d4ca766            Control summary-Main-save   

                              Resource       Complete Timestamp  \
143184                          3d952e  2015-06-30 08:03:34.000   
1494673  Document processing automaton  2016-08-10 13:36:37.781   
1495362                         167893  2016-12-14 17:07:22.000   

         (case) amount_applied0  (case) amount_applied1  \
143184                 56564.75                56564.75   
1494673                34769.12                34794.71   
1495362                34769.12                34794.71   

         (case) amount_applied2  (case) amount_applied3  (case) applicant  \
143184                 56482.47                56550.60  748281cc09b65ef9   
1494673                35254.05                35253.72  a5d52351e3091a8d   
1495362                35254.05                35253.72  a5d52351e3091a8d   

         (case) area  (case) basic payment  (case) cross_compliance  \
143184      187.7858                  True                      0.0   
1494673     107.9741                  True                      0.0   
1495362     107.9741                  True                      0.0   

        (case) department  (case) greening  (case) number_parcels  \
143184                 6b             True                     57   
1494673                d4             True                     59   
1495362                d4             True                     59   

         (case) payment_actual0  (case) payment_actual1  \
143184                 54786.25                    0.00   
1494673                29492.92                 5273.58   
1495362                29492.92                 5273.58   

         (case) payment_actual2  (case) payment_actual3  (case) penalty_ABP  \
143184                  -252.02                     0.0                True   
1494673                 -376.22                  -378.3               False   
1495362                 -376.22                  -378.3               False   

         (case) penalty_AGP  (case) penalty_AJLP  (case) penalty_AUVP  \
143184                 True                False                 True   
1494673               False                False                False   
1495362               False                False                False   

         (case) penalty_AVBP  (case) penalty_AVGP  (case) penalty_AVJLP  \
143184                 False                False                 False   
1494673                False                False                 False   
1495362                False                False                 False   

         (case) penalty_AVUVP  (case) penalty_B16  (case) penalty_B2  \
143184                  False               False               True   
1494673                 False               False              False   
1495362                 False               False              False   

         (case) penalty_B3  (case) penalty_B4  (case) penalty_B5  \
143184               False              False              False   
1494673              False              False               True   
1495362              False              False               True   

         (case) penalty_B5F  (case) penalty_B6  (case) penalty_BGK  \
143184                False              False               False   
1494673               False              False               False   
1495362               False              False               False   

         (case) penalty_BGKV  (case) penalty_BGP  (case) penalty_C16  \
143184                 False               False               False   
1494673                False               False               False   
1495362                False               False               False   

         (case) penalty_C4  (case) penalty_C9  (case) penalty_CC  \
143184               False  

### Case lifecycle

In [9]:
len(set(elog['Case ID']))

43809

So there are 43809 cases present in the event log. Let's see how many of these cases are actually useful to use for training/testing/cross-validation.

In [11]:
elog.loc[elog['Activity'] == 'Payment application-Application-decide'].shape

(65437, 89)

In [12]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID']))

43793

We see that there are only 43809 - 43793 = 16 cases that do not have the the 'Payment application-Application-decide' activity in their lifecycle (as far as can be deduced from the event logs)

Let's drop these 16 cases since they are incomplete in this context and are not that significant when you consider the number of cases at hand.

In [15]:
to_be_dropped_cases = set(elog['Case ID']).difference(set(set(elog.loc[elog['Activity'] == 'Payment application-Application-decide']['Case ID'])))

In [23]:
elog = elog[~elog['Case ID'].isin(to_be_dropped_cases)]

But since we want to be able to detect the undesired outcomes before the 'Payment application-Application-decide' the choice is made to consider the event that proceeds this activity in almost all the cases: 'Payment application-Application-finish editing'.

In [14]:
len(set(elog.loc[elog['Activity'] == 'Payment application-Application-finish editing']['Case ID']))

43795

In [ ]:
e